Improve rustdoc error for failed intra-doc link resolution

camelid opened the pull request:
The previous error was confusing since it made it sound like you can't
link to items that are defined outside the current module.

Also suggested importing the item.

r? @jyn514

rust-highfive commented:
Some changes occurred in intra-doc-links.

cc @jyn514

camelid commented on src/librustdoc/passes/collect_intra_doc_links.rs:
```diff
@@ -1575,17 +1575,17 @@ fn resolution_failure(
                         } else {
                             diag.note(&note);
                         }
+                        diag.help(&format!("did you mean to import `{}`?", unresolved));
```
I should probably remove this part since it's not super helpful and could be confusing, but I figured that I would post it so you could see what it looks like.

jyn514 commented on src/librustdoc/passes/collect_intra_doc_links.rs:
```diff
@@ -1575,17 +1575,17 @@ fn resolution_failure(
                         } else {
                             diag.note(&note);
                         }
+                        diag.help(&format!("did you mean to import `{}`?", unresolved));
```
I don't think this is very helpful is we don't suggest where to import it _from_. It's just as likely this was a typo or a renamed method.

camelid replied:
See my comment [above](https://github.com/rust-lang/rust/pull/77469#discussion_r499061527) :)

I preemptively removed it locally; I'll amend the commit.

camelid replied:
Actually I'll make a new commit, and squash at the end so it's easy to review.

jyn514 commented on src/librustdoc/passes/collect_intra_doc_links.rs:
```diff
@@ -1575,17 +1575,17 @@ fn resolution_failure(
                         let note = format!(
-                            "the module `{}` contains no item named `{}`",
-                            module_name, unresolved
+                            "there is no item named `{}` in scope",
```
This should only change if there is no intermediate path: `[x]`. If we have `[std::io::nope]` then this is actively misleading. The way to tell that is `partial_res.is_none()`.

camelid replied:
I have:

```rust
                        let note = if partial_res.is_none() {
                            format!("there is no item named `{}` in scope", unresolved)
                        } else {
                            format!("not in scope")
                        };
```

Which produces this output:

```
error: unresolved link to `path::to::nonexistent::module`
  --> $DIR/intra-link-errors.rs:7:6
   |
LL | /// [path::to::nonexistent::module]
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ there is no item named `path` in scope
   |
note: the lint level is defined here
  --> $DIR/intra-link-errors.rs:1:9
   |
LL | #![deny(broken_intra_doc_links)]
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: unresolved link to `path::to::nonexistent::macro`
  --> $DIR/intra-link-errors.rs:11:6
   |
LL | /// [path::to::nonexistent::macro!]
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ there is no item named `path` in scope

error: unresolved link to `path::to::nonexistent::type`
  --> $DIR/intra-link-errors.rs:15:6
   |
LL | /// [type@path::to::nonexistent::type]
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ there is no item named `path` in scope

error: unresolved link to `std::io::not::here`
  --> $DIR/intra-link-errors.rs:19:6
   |
LL | /// [std::io::not::here]
   |      ^^^^^^^^^^^^^^^^^^ not in scope

error: unresolved link to `std::io::not::here`
  --> $DIR/intra-link-errors.rs:23:6
   |
LL | /// [type@std::io::not::here]
   |      ^^^^^^^^^^^^^^^^^^^^^^^ not in scope
```

What should the messages be?

jyn514 replied:
The ones for `path` look fine. I would keep the others the way they are now: 'no item named `not` in module `io`'.

camelid commented on src/librustdoc/passes/collect_intra_doc_links.rs:
```diff
@@ -1576,22 +1576,28 @@ fn resolution_failure(
+                        let note = if partial_res.is_some() {
+                            let module_name = collector.cx.tcx.item_name(module);
+                            format!(
+                                "the module `{}` contains no item named `{}`",
```
Should I change this error to something like ``there is no item named `{}` in the module `{}` `` to be consistent with the other error?

jyn514 replied:
```suggestion
                                "no item named `{}` in module`{}`",
```
seems concise and matches the other error.

camelid commented on src/librustdoc/passes/collect_intra_doc_links.rs:
```diff
@@ -1576,22 +1576,28 @@ fn resolution_failure(
+                                module_name, unresolved
+                            )
+                        } else {
+                            format!("there is no item named `{}` in scope", unresolved)
```
This is a little wordy; should I change it to this? I'm not sure if it's less clear though.

```suggestion
                            format!("`{}` is not in scope", unresolved)
```

jyn514 replied:
I think
```suggestion
                            format!("no item named `{}` in scope", unresolved)
```
would be equally clear and less verbose. 'x is not in scope' works but seems a little _too_ concise, IDK. I'd be fine with either.

camelid replied:
I at first went for "{} is not in scope" but then I realized yours was clearer :)

The shorter one "looks" nicer, but it makes it sound like there definitely is something named "{}", you just need to import it, when in fact it's possible it doesn't exist. Maybe when we get access to `rustc_resolve`'s fancy errors we can change the message based on whether there is something named "{}" that you could import.

camelid commented:
Huh, GitHub's labels have been acting up recently; it shows that I removed `T-rustdoc` after you added it, even though I added it then. I guess there was a race condition? They should use Rust on their backend! ;)

jyn514 commented:
Yeah, this looks a lot better. r=me with CI passing.

jyn514 commented:
@bors r+

Thanks for working on this!

bors commented:
:pushpin: Commit aa9b718cf0d4157a81c5dcc8d59f2c4ed96b801e has been approved by `jyn514`

<!-- @bors r=jyn514 aa9b718cf0d4157a81c5dcc8d59f2c4ed96b801e -->
<!-- homu: {"type":"Approved","sha":"aa9b718cf0d4157a81c5dcc8d59f2c4ed96b801e","approver":"jyn514"} -->
